### PR TITLE
feat: additional x/leverage queries

### PR DESCRIPTION
**Pull request by toteki**
## Description

Adds more queries (and CLI support) for x/leverage

Progress towards #93 and #94

----

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [x] added appropriate labels to the PR
- [ ] added `!` to the type prefix if API or client breaking change
- [x] targeted the correct branch (see [PR Targeting](https://github.com/umee-network/umee/blob/main/CONTRIBUTING.md#pr-targeting))
- [x] provided a link to the relevant issue or specification
- [ ] added a changelog entry to `CHANGELOG.md`
- [x] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [x] reviewed "Files changed" and left comments if necessary
- [x] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: x/leverage/client/cli/query.go [hunk 1]
```diff
@@ -30,6 +30,9 @@ func GetQueryCmd(queryRoute string) *cobra.Command {
 		GetCmdQueryParams(),
 		GetCmdQueryBorrowed(),
 		GetCmdQueryReserveAmount(),
+		GetCmdQueryCollateral(),
```
Comment (alexanderbez) [inline x/leverage/client/cli/query.go:33]: Can we add integration tests for these please?

**Review by alexanderbez — CHANGES_REQUESTED** [review]

File: x/leverage/client/cli/query.go [hunk 2]
```diff
@@ -162,3 +165,107 @@ func GetCmdQueryReserveAmount() *cobra.Command {
...
+	}
+
+	flags.AddQueryFlagsToCmd(cmd)
+
+	return cmd
+}
+
+// GetCmdQueryCollateralSetting returns a CLI command handler to query for the amount of
+// total collateral tokens for a given address.
+func GetCmdQueryCollateralSetting() *cobra.Command {
```
Comment (toteki) [inline x/leverage/client/cli/query.go:205]:
```suggestion
// GetCmdQueryCollateral returns a CLI command handler to query for the amount of
// total collateral tokens for a given address.
func GetCmdQueryCollateral() *cobra.Command {
```

File: x/leverage/client/cli/query.go [hunk 3]
```diff
@@ -162,3 +165,107 @@ func GetCmdQueryReserveAmount() *cobra.Command {
 
 	return cmd
 }
+
+// GetCmdQueryCollateral returns a CLI command handler to query for the amount of
+// total collateral tokens for a given address.
+func GetCmdQueryCollateral() *cobra.Command {
```
Comment (toteki) [inline x/leverage/client/cli/query.go:171]:
```suggestion
// GetCmdQueryCollateralSetting returns a CLI command handler to query for the collateral
// setting of a specific token denom for an address.
func GetCmdQueryCollateralSetting() *cobra.Command {
```

**Comment by toteki** [comment]
R4R again - note that the tests in `client/tests/suite.go` are still at a bare minimum - the next PR will flesh them out (table-driven, additional setup, etc).

File: x/leverage/keeper/grpc_query.go [hunk 4]
```diff
@@ -109,3 +109,134 @@ func (q Querier) ReserveAmount(
...
+	setting := q.Keeper.GetCollateralSetting(ctx, borrower, req.Denom)
+
+	return &types.QueryCollateralSettingResponse{Enabled: setting}, nil
+}
+
+func (q Querier) Collateral(
+	goCtx context.Context,
+	req *types.QueryCollateralRequest,
+) (*types.QueryCollateralResponse, error) {
+
```
Comment (adamewozniak) [inline x/leverage/keeper/grpc_query.go:144]:
```suggestion
```

File: x/leverage/keeper/grpc_query.go [hunk 5]
```diff
@@ -109,3 +109,134 @@ func (q Querier) ReserveAmount(
 
 	return &types.QueryReserveAmountResponse{Amount: amount}, nil
 }
+
+func (q Querier) CollateralSetting(
+	goCtx context.Context,
+	req *types.QueryCollateralSettingRequest,
+) (*types.QueryCollateralSettingResponse, error) {
+
```
Comment (adamewozniak) [inline x/leverage/keeper/grpc_query.go:117]:
```suggestion
```

File: x/leverage/keeper/grpc_query.go [hunk 6]
```diff
@@ -109,3 +109,134 @@ func (q Querier) ReserveAmount(
...
+	token := q.Keeper.GetCollateralAmount(ctx, borrower, req.Denom)
+
+	return &types.QueryCollateralResponse{Collateral: sdk.NewCoins(token)}, nil
+}
+
+func (q Querier) ExchangeRate(
+	goCtx context.Context,
+	req *types.QueryExchangeRateRequest,
+) (*types.QueryExchangeRateResponse, error) {
+
```
Comment (adamewozniak) [inline x/leverage/keeper/grpc_query.go:174]: same for the rest of them
```suggestion
```

File: x/leverage/client/tests/cli_test.go [hunk 7]
```diff
@@ -16,3 +16,10 @@ func TestIntegrationTestSuite(t *testing.T) {
 	cfg.NumValidators = 2
 	suite.Run(t, NewIntegrationTestSuite(cfg))
 }
+
+func (s *IntegrationTestSuite) TearDownSuite() {
```
Comment (alexanderbez) [inline x/leverage/client/tests/cli_test.go:20]: Good catch. Thanks!

File: x/leverage/keeper/collateral.go [hunk 8]
```diff
@@ -77,3 +77,10 @@ func (k Keeper) GetBorrowerCollateral(ctx sdk.Context, borrowerAddr sdk.AccAddre
 	totalCollateral.Sort()
 	return totalCollateral
 }
+
+// GetEligibleLiquidationTargets returns a list of borrower addresses eligible for liquidation
+func (k Keeper) GetEligibleLiquidationTargets(ctx sdk.Context) ([]sdk.AccAddress, error) {
```
Comment (alexanderbez) [inline x/leverage/keeper/collateral.go:82]: Do we actually need this to be defined in the state-machine? Can the tool not just query on-chain state to figure out the targets?

Granted, we could just build the query for them too...

**Review by alexanderbez — APPROVED** [review]

**Pull request merged**